### ISSUE TITLE
Investigation of Possible Latency Issue Due to Aggressive Pipelining Issue

### ISSUE THREAD

**Issue by rkarthick** [StackExchange/StackExchange.Redis#2462]
We're currently experiencing a latency issue (high p99) in our production environment. The investigations have so far excluded server issues, network issues, and client thread contention as potential culprits.  Our application architecture involves the use of an Envoy proxy, with all Redis commands flowing through this proxy to reach the cluster. We are now exploring the possibility that aggressive pipelining might be contributing to the observed latency, potentially due to head-of-line blocking.

In order to delve deeper into this hypothesis, I am trying to examine the total number of commands being dispatched simultaneously via pipelining. The Send_PauseWriterThreshold is configured to be [at least 0.5MB](https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/SocketManager.cs#L88), while the MinimumSegmentSize is set at a minimum of 8KB. Given these settings, does this imply that if the Redis commands are only a few bytes in size, we should expect to see at least 8KB divided by the size of the Redis command worth of commands being "batched" together? Is this a correct perspective on pipelining?

I'm contemplating integrating a [string split here](https://github.com/mgravell/Pipelines.Sockets.Unofficial/blob/main/src/Pipelines.Sockets.Unofficial/SocketConnection.Send.cs#L65) into our private fork to ascertain the quantity of commands sent concurrently. However, before proceeding with this approach, I thought it would be wise to inquire here if there are other effective methods to achieve this goal.

**Comment by NickCraver** [comment]
My understanding is `MinimumSegmentSize` is about buffer sizes and not required batching to the pipe. Think of it this way: average MTU sizing is 1.4-1.5K, so an 8KB batch really doesn't make any sense for any socket-based usages unless they're on jumbo frames (which is still fairly rare in the grand scheme of things...so we'd be talking fragmentation for everyone else). I think the question being asked here is "how is Nagle's in play with the default settings in Pipelines.Sockets.Unofficial?" Is that a decent description of what you're looking for?

Definitely need @mgravell to chime in here on intended behavior but I wanted to ask: on the debugging front: have you gotten packet captures? That'd be the quickest definitive way to see what's being sent on which timings at the socket level then we can trace back up to see where you're hitting issues. Comparing timings to captures from Envoy as well would be equally enlightening to see what latency is in play there.

FWIW, one trick I tend to use here is: generate a ton of GUIDs (millions), either on the fly or ahead of time depending on how fast we need to be, and send those commands. It can be simple string gets with all key misses, the result doesn't matter unless your result payloads are what's causing delay in the first place (e.g. incoming bandwidth starvation under load). The GUID approach is because it makes it much easier to trace _a command_ through multiple packet captures from multiple machines, as well as see how many GUIDs are in each.

**Comment by rkarthick** [comment]
Yeah, understanding the interplay between Nagle's algorithm and Pipelines.Sockets.Unofficial will be useful. More so, gaining deeper insight into how many commands are grouped together into a single network packet, knobs that we can tune for grouping, would be valuable to rule out pipelining as a variable in our investigation. On another note, it's interesting that the p99 latency observed by Envoy to the Redis cluster is an order of magnitude (10x) less than the overall end-to-end latency observed from the client. As for packet captures, I haven't explored that avenue yet. Thank you for the suggestion. We also _may_ have jumbo packets enabled in our infra. I will circle back with my findings.

**Comment by mgravell** [comment]
Thanks for the report. StackExchange.Redis **3.0** ships a substantial rewrite of the low-level IO core — a new RESPite-based stack that replaces `Pipelines.Sockets.Unofficial` and the old backlog/write loop — and we're confident this issue is addressed by that work.

Closing as resolved in 3.0. If you can still reproduce on 3.0.x, please comment and we'll reopen and investigate with fresh diagnostics.
